File: 79.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        w = len(word)
        used = [[False] * n for _ in range(m)]

        def backtrack(i, j, ptr):
            if ptr == w - 1:
                return True
            used[i][j] = True
            for (di, dj) in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                ii = i + di
                jj = j + dj
                if 0 <= ii < m and 0 <= jj < n and board[ii][jj] == word[ptr + 1] and not used[ii][jj]:
                    if backtrack(ii, jj, ptr + 1):
                        return True
            used[i][j] = False
            return False

        for state in ((i, j) for i in range(m) for j in range(n) if board[i][j] == word[0]):
            if backtrack(*state, 0):
                return True
        return False

File: sol/79.cpp
struct Solution {
    bool backtrack(
            int i, int j, int ptr, 
            const vector<vector<char>> &board, 
            const string &word, vector<vector<bool>> &used){
        if (i < 0 || i >= board.size() || j < 0 || j >= board.front().size() || board[i][j] != word[ptr] || used[i][j]){
            return false;
        }
        if (ptr == word.size() - 1){
            return true;
        }
        used[i][j] = true;
        ++ptr;
        bool res = backtrack(i - 1, j, ptr, board, word, used)
                || backtrack(i + 1, j, ptr, board, word, used)
                || backtrack(i, j - 1, ptr, board, word, used)
                || backtrack(i, j + 1, ptr, board, word, used);
        used[i][j] = false;
        return res;
    }

    bool exist(vector<vector<char>>& board, string &word) {
        vector<vector<bool>> used(board.size(), vector<bool>(board.front().size(), false));
        for (int i = 0; i < board.size(); ++i){
            for (int j = 0; j < board.front().size(); ++j){
                if (backtrack(i, j, 0, board, word, used)){
                    return true;
                }
            }
        }
        return false;
    }
};
